### PR TITLE
fix: add item to workqueue with delay when syncFailed

### PR DESCRIPTION
**Pull request by gxthrj**
Fix: #59

### CODE REVIEW

File: pkg/ingress/controller/apisix_service.go [hunk 1]
```diff
@@ -89,16 +95,15 @@ func (c *ApisixServiceController) processNextWorkItem() bool {
...
 	err := func(obj interface{}) error {
 		defer c.workqueue.Done(obj)
-		var key string
+		var sqo *ServiceQueueObj
 		var ok bool
-
-		if key, ok = obj.(string); !ok {
+		if sqo, ok = obj.(*ServiceQueueObj); !ok {
 			c.workqueue.Forget(obj)
 			return fmt.Errorf("expected string in workqueue but got %#v", obj)
```
Comment (tokers) [inline pkg/ingress/controller/apisix_service.go:102]: Typo, should be "expected ServiceQueueObj object in workqueue but got %#v".

File: pkg/ingress/controller/apisix_route.go [hunk 2]
```diff
@@ -100,14 +100,15 @@ func (c *ApisixRouteController) updateFunc(oldObj, newObj interface{}) {
 }
 
 func (c *ApisixRouteController) deleteFunc(obj interface{}) {
+	oldRoute := obj.(cache.DeletedFinalStateUnknown).Obj.(*api6V1.ApisixRoute)
```
Comment (tokers) [inline pkg/ingress/controller/apisix_route.go:103]: Reference to istio: https://github.com/istio/istio/blob/18b3d35de0523862a389553b50b13ee4e73f4530/pilot/pkg/serviceregistry/kube/controller/namespacecontroller.go#L160

File: pkg/ingress/controller/apisix_route.go [hunk 3]
```diff
@@ -211,18 +212,15 @@ func (c *ApisixRouteController) sync(rqo *RouteQueueObj) error {
...
-		}
-		runtime.HandleError(fmt.Errorf("failed to list apisixRoute %s/%s", key, err.Error()))
-		return err
-	}
 	switch {
 	case rqo.Ope == UPDATE:
+		apisixIngressRoute, err := c.apisixRouteList.ApisixRoutes(namespace).Get(name)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				log.Errorf("apisixRoute %s is removed", key)
```
Comment (tokers) [inline pkg/ingress/controller/apisix_route.go:220]: If the apisixRoute is removed, we don't need to retry, so just return `nil`.

File: pkg/ingress/controller/apisix_route.go [hunk 4]
```diff
@@ -232,10 +230,15 @@ func (c *ApisixRouteController) sync(rqo *RouteQueueObj) error {
 		rc := &state.RouteCompare{OldRoutes: oldRoutes, NewRoutes: newRoutes}
 		return rc.Sync()
 	case rqo.Ope == DELETE:
-		apisixRoute := apisix.ApisixRoute(*apisixIngressRoute)
+		apisixIngressRoute, _ := c.apisixRouteList.ApisixRoutes(namespace).Get(name)
+		if apisixIngressRoute != nil && apisixIngressRoute.ResourceVersion > rqo.OldObj.ResourceVersion {
```
Comment (tokers) [inline pkg/ingress/controller/apisix_route.go:234]: We'd better to log it, it's somewhat confusing (if we don't know the retry knowledge background).

File: pkg/ingress/controller/apisix_route.go [hunk 5]
```diff
@@ -144,6 +144,7 @@ func (c *ApisixRouteController) processNextWorkItem() bool {
 		}
 		// 在syncHandler中处理业务
 		if err := c.syncHandler(rqo); err != nil {
+			c.workqueue.AddRateLimited(obj)
```
[the remark this answers is not in the document]
Comment (gxthrj) [reply pkg/ingress/controller/apisix_route.go:146]: Done

File: pkg/ingress/controller/apisix_service.go (hunk 1 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_service.go]: Done

File: pkg/ingress/controller/apisix_route.go (hunk 2 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_route.go]: Done

File: pkg/ingress/controller/apisix_route.go (hunk 3 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_route.go:230]: Yes

File: pkg/ingress/controller/apisix_route.go (hunk 4 above)
Comment (gxthrj) [reply pkg/ingress/controller/apisix_route.go:245]: Done, add a warn level log

**Review by tokers — APPROVED** [review]

**Pull request merged**